### ISSUE TITLE
OSX: Signed app

### ISSUE THREAD

**Issue by Floppy** [theodi/comma-chameleon#93]
Currently the app isn't signed, so gets flagged under the unknown developer security thing.

**Comment by Stephen-Gates** [comment]
+1 
Tried making Comma Chameleon my default tool to open CSV files (1) for a demo (2) because Excel mangles data, but...

<img width="432" alt="screenshot 2016-09-14 07 07 01" src="https://cloud.githubusercontent.com/assets/9379524/18491582/647c6760-7a4a-11e6-9a33-75e8382eee81.png">

**Comment by drchriscole** [comment]
Saw this tool as highly recommended, but can't even open it.

**Comment by Floppy** [comment]
@drchriscole you can right click then option-click on open to enable the app, but yes, this is something we need to fix properly.

**Comment by Floppy** [comment]
Looks like https://www.npmjs.com/package/electron-osx-sign might be the thing to use

**Comment by Floppy** [comment]
@davetaz am I right in thinking we already have an apple developer license for the ODI because of the mobile apps?

**Comment by davetaz** [comment]
Correct, you should be an admin on it I think. Check your apple id at developer.apple.com. You should then be able to provision an app profile and somehow get the certificates needed for solving this.

**Comment by Floppy** [comment]
@davetaz brilliant, thanks.

**Comment by pezholio** [comment]
Just clocked this. Just one issue to flag, we might need to investigate how the releases are built as I **think** you need to build on OSX if you want to sign an app for OSX. I'm pretty sure Travis has some OSX images though

**Comment by Floppy** [comment]
Yep, need to build (or at least sign) on OSX. Hadn't considered doing travis build on OSX, that's an interesting possibility; I was thinking it would have to be a manual process (which obviously would suck). I'll take a look.

**Comment by Floppy** [comment]
`electron-packager` (which we use already) uses `electron-osx-sign` so there's no need to change that. Just need to be on the right OS and have the certificate.

**Comment by pezholio** [comment]
Ace 👍

**Comment by drchriscole** [comment]
@Floppy thanks. The right-click allowed me to run it. Nice trick! Thanks.

**Comment by quadrophobiac** [comment]
So I cannot duplicate the `unidentified developer` warning message - can someone who had this problem previously confirm if their `Security & Privacy` settings were set like so?
![screen shot 2017-06-02 at 12 54 26](https://cloud.githubusercontent.com/assets/1039017/26724686/a604b06e-4792-11e7-9145-71baa8e50c84.png)

**Comment by Stephen-Gates** [comment]
To re-create:

- download app
- move to application directory
- go to finder, right click on .csv file
- choose open with Comma Chameleon

<img width="419" alt="screenshot 2017-06-02 22 15 56" src="https://cloud.githubusercontent.com/assets/9379524/26725319/561c5fda-47e1-11e7-98ab-c24389d380c9.png">

**Comment by quadrophobiac** [comment]
Thanks @Stephen-Gates, for whatever reason I can't replicate it exactly **but** I do have something that will let me check if my signing remedies it

**Comment by quadrophobiac** [comment]
I've used [RB App Checker Lite](https://itunes.apple.com/gb/app/rb-app-checker-lite/id519421117?mt=12) to verify if the app is being signed correctly. For comparison here is the current package as generated by the gulp script, then that same package signed with `electron-osx-sign` 
![gulp-generated-comma-chameleon](https://cloud.githubusercontent.com/assets/1039017/26783302/59cc4dce-49ef-11e7-9042-749d25dd5624.png)
![gulp-generated-comma-chameleon-after-sign](https://cloud.githubusercontent.com/assets/1039017/26783303/5b0f6c02-49ef-11e7-96bc-400ac2cbf2d7.png)

The above signature differences are replicated in a more manual signing flow where `electron-packager` and `electron-sign` are utilised 

While the correct signatures have been applied I am seeing no difference in behaviour in terms of being the default opening app. I have updated releases - only MacOS `Comma Chameleon-darwin-x64` has been signed - from [this related branch](https://github.com/theodi/comma-chameleon/tree/change-package-favicon) if anyone here would care to test if any behaviour has changed. For now the same build flow has been utilised to update the releases page by a minor increment. https://github.com/theodi/comma-chameleon/releases/tag/0.5.2

Given I expect that the signing won't resolve the problems indicated thus far I think the issues in this ticket need separating out, one for app showing up as a Signed Identified App, and the other to relate to setting electron app as default for opening CSV files. There is also a question of how the current utilisation of csvlint should work - as a binary it should also be signed.

It may be necessary to change the build process wholesale by switching to [electron-builder](https://github.com/electron-userland/electron-builder). It seems a lot of people defer to this after encountering many of the same issues I encountered trying to get this set up. It also has support for yarn, and may play well with dependabot (cc @floppy)

**Comment by Stephen-Gates** [comment]
@quadrophobiac 

Downloaded App
<img width="767" alt="screenshot 2017-06-05 22 40 30" src="https://cloud.githubusercontent.com/assets/9379524/26784368/8b43dcae-4a40-11e7-94a6-a5c42989cde0.png">

Used Finder to open a CSV with Comma Chameleon
<img width="711" alt="screenshot 2017-06-05 22 42 30" src="https://cloud.githubusercontent.com/assets/9379524/26784398/b4f0b928-4a40-11e7-8712-3654e345c6ba.png">

Didn't like it
<img width="421" alt="screenshot 2017-06-05 22 42 55" src="https://cloud.githubusercontent.com/assets/9379524/26784411/c67ce356-4a40-11e7-936f-48ba4c9af540.png">

Adjusted settings and still couldn't open csv file from finder but could open using File > Open in Comma Chameleon

**Comment by quadrophobiac** [comment]
Yeah that's the same issue I encountered. I've researched this and I don't see that there's anything else to be done in terms of signing.

**Comment by Stephen-Gates** [comment]
So is the solution to submit to the App Store?

**Comment by quadrophobiac** [comment]
In short I don't know at present - my gut says no, because mac app store and non-app store distribution are both supposed to be supported. Trouble is most guides assume that xcode is being used to sign apps

**Comment by quadrophobiac** [comment]
@Stephen-Gates can you try these steps 

1. download the Comma Chameleon link I emailed you (it's a way of mocking the quarantine steps [here](https://developer.apple.com/library/content/documentation/IDEs/Conceptual/AppDistributionGuide/DistributingApplicationsOutside/DistributingApplicationsOutside.html#//apple_ref/doc/uid/TP40012582-CH12-SW1)) 
2. Set your gatekeeper to `Allow apps downloaded from: App Store`
You should see this alert
![screen shot 2017-06-05 at 15 45 53](https://cloud.githubusercontent.com/assets/1039017/26789004/d949c628-4a06-11e7-9b7a-c9a8afbcf162.png)
![screen shot 2017-06-05 at 15 46 35](https://cloud.githubusercontent.com/assets/1039017/26789010/deaef6ec-4a06-11e7-814f-267e70628891.png)
Change the Gatekeeper settings to `Allow apps downloaded from: App Store and identified developers`
You should get 
![screen shot 2017-06-05 at 15 52 33](https://cloud.githubusercontent.com/assets/1039017/26789121/38ebe9da-4a07-11e7-9d81-8f7385172b36.png)
and thereafter the app should open fine.

None of this resolves the 'set as default CSV opening app' but I think it confirms that the app is displaying the requisite behaviour for a developer signed app. 

Given the restrictions that Sierra brought to Gatekeeper working towards a Mac App store distribution is definitely the best solution to this problem overall. It's also possible that sandboxing might be needed, and doing that is part of the Mac App Store distro process. So if you can confirm that you see the above Stephen I'll move to close this issue

**Comment by Stephen-Gates** [comment]
Yep. I get the some as above @quadrophobiac

**Comment by quadrophobiac** [comment]
In response to the initial ticket
>Currently the app isn't signed, so gets flagged under the unknown developer security thing.
I think this is a separate issue to the open CSV by default. It's unclear from electron documentation how to accomplish this without submitting to the App Store. Creating a sandboxed app with the correct manifest details may resolve it. I noticed a closed issue - https://github.com/theodi/comma-chameleon/issues/75-  on the same request, may warrant reopening it in light of this lack of functionality. 
In the meantime including the 'right click workaround' within the readme seems the most expedient solution to the problem
